spec/integration/game_spec.rb

```diff
@@ -0,0 +1,33 @@
+require 'rspec'
+require 'rspec/core'
+require_relative '../../lib/gamestate.rb'
+
+describe Gamestate do
+  let(:gamestate) { Gamestate.new(@castle_data) }
+
+  context "winning conditions" do
+    before do
+      file = File.read('./spec/integration/config/game-layout-win.json')
+      @castle_data = JSON.parse(file)
+    end
+
+    it "validates the player can win a game" do
+      allow(STDIN).to receive(:gets).and_return("fight")
+      expect(gamestate).to receive(:win)
+      gamestate.play
+    end
+  end
+
+  context "losing conditions" do
+    before do
+      file = File.read('./spec/integration/config/game-layout-lose.json')
+      @castle_data = JSON.parse(file)
+    end
+
+    it "validates the player can lose a game" do
+      allow(STDIN).to receive(:gets).and_return("fight")
+      expect(gamestate).to receive(:ask_player_to_try_again)
+      gamestate.play
+    end
+  end
+end
```

Comment:
Several issues here:

- Idiomatic rspec uses let variables, not @attributes.
- There is much duplication both within the test and between the test and Gamestate.
- Stubbing the output methods means they're not integration-tested.

To address this, I'd write end-to-end integration tests which enter the game file name as well as subsequent commands, and assert that the game prints the expected output rather than stubbing.

If you did that the other aspects of code that I'm complaining about would go away. But be aware that, if you wanted to keep the current test approach but just clean up the code issues, you could make the attribute a let variable and you could move the duplicated code in the before blocks to the top level.

Reply:
Note that you don't need such long test files for these tests. Test data should be as small as possible while still supporting the test that you're writing. It will be easier for others to understand and easier for you to test precisely.

Reply:
I will admit after some trial and error that I should have followed up sooner on this comment to either: 

- get further clarity
- take advantage of a learning moment 

I acknowledge that I went a layer deeper than I should have with the filename but the rest I'm interested to learn more. This is important to me as integration testing is crucial. 

I cleaned it up just a little but I have a feeling it's not quite right if stubbing wasn't the way to go. Does this mean I should have used mocks? I was feeling good after watching the game simulate a playthrough both with a win and loss result. I will push what I have for now and follow up with Savannah / you later.